system/lib/wasmfs/file.h

```diff
@@ -8,37 +8,133 @@
 
 #pragma once
 
+#include <assert.h>
 #include <emscripten/html5.h>
 #include <mutex>
+#include <unordered_map>
 #include <vector>
 #include <wasi/api.h>
 
 namespace wasmfs {
 
-class File {
-  // TODO: Add other File properties later.
+class File : public std::enable_shared_from_this<File> {
+
+public:
+  enum FileKind { DataFileKind = 0, DirectoryKind, SymlinkKind };
+
+  // curr->is<Directory>()
+  template <class T> bool is() const {
+    static_assert(std::is_base_of<File, T>::value, "File is not a base of destination type T");
+    return int(kind) == int(T::SpecificFileKind);
+  }
+
+  template <class T> T* dynCast() {
+    static_assert(std::is_base_of<File, T>::value, "File is not a base of destination type T");
+    return int(kind) == int(T::SpecificFileKind) ? (T*)this : nullptr;
+  }
+
+  template <class T> const T* dynCast() const {
+    static_assert(std::is_base_of<File, T>::value, "File is not a base of destination type T");
+    return int(kind) == int(T::SpecificFileKind) ? (const T*)this : nullptr;
+  }
```

Comment:
Are the `const` overloads used anywhere? If not, we can probably remove them for now.